### PR TITLE
Add cancellation token support to Subscribers

### PR DESCRIPTION
**Pull request by matt-psaltis**
Relates to: https://github.com/dotnetcore/CAP/issues/907

Wanted to get your thoughts on this currently. I've added an example integration test as well as some additional unit tests for the subscriber resolution when cancellation tokens are in play. There's a couple of signature changes that would be potentially breaking so I'd like to get your thoughts on those please?

### CODE REVIEW

File: src/DotNetCore.CAP/Internal/ICapPublisher.Default.cs [hunk 1]
```diff
@@ -38,6 +38,11 @@ public CapPublisher(IServiceProvider service)
 
         public AsyncLocal<ICapTransaction> Transaction { get; }
 
+        public void Start(CancellationToken stoppingToken)
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP/Internal/ICapPublisher.Default.cs:41]: Should it be removed?  I think the prevent Start method in CapPublisher looks a bit strange?  and the method is not called

File: src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs [hunk 2]
```diff
@@ -36,7 +36,7 @@ public async Task<ConsumerExecutedResult> InvokeAsync(ConsumerContext context, C
             cancellationToken.ThrowIfCancellationRequested();
 
             var methodInfo = context.ConsumerDescriptor.MethodInfo;
-            var reflectedType = methodInfo.ReflectedType.Name;
+            var reflectedType = methodInfo.ReflectedType.TypeHandle.Value;
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs:39]: If we use `ReflectedType.TypeHandle.Value`, as described in #908, do we still need `"{methodInfo.Module.Name}_{reflectedType}_{methodInfo.MetadataToken}"` to ensure that the key is unique?

**Review by yang-xiaodong — COMMENTED** [review]
Hello @matt-psaltis,

Thank you very much for your PR,  I reviewed and found two issues, please confirm it

**Comment by yang-xiaodong** [comment]
Hello @matt-psaltis , Any further response on this PR?

File: src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs (hunk 2 above)
Comment (matt-psaltis) [reply src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs:39]: Hey got slammed with work this week. I was trying to keep the PRs separate so I was planning to do this once I had some test coverage in place in a subsequent PR. happy to amalgamate tho if that's easier.

File: src/DotNetCore.CAP/Internal/ICapPublisher.Default.cs (hunk 1 above)
Comment (matt-psaltis) [reply src/DotNetCore.CAP/Internal/ICapPublisher.Default.cs:41]: It does look strange on the publisher side agreed. I need to come back to this and reassess. It didn't quite fit the current structure.

**Comment by matt-psaltis** [comment]
> Hello @matt-psaltis , Any further response on this PR?

@yang-xiaodong I've pushed some updates and additional tests for the cache key problem from #908 Would love your thoughts when you can please.

**Comment by yang-xiaodong** [comment]
Hello @matt-psaltis ,

Since the object that implements `IProcessingServer` will be started in bootstrap so what do you think  about making the `IDispatcher`  inherit the `IProcessingServer` ? 

https://github.com/dotnetcore/CAP/blob/master/src/DotNetCore.CAP/Internal/IBootstrapper.Default.cs#L33

**Comment by matt-psaltis** [comment]
> Hello @matt-psaltis ,
> 
> Since the object that implements `IProcessingServer` will be started in `Bootstrapper` class so what do you think about making the `IDispatcher` inherit the `IProcessingServer` ?
> 
> https://github.com/dotnetcore/CAP/blob/master/src/DotNetCore.CAP/CAP.ServiceCollectionExtensions.cs#L47
> https://github.com/dotnetcore/CAP/blob/master/src/DotNetCore.CAP/Internal/IBootstrapper.Default.cs#L33

Awesome! I actually did go down that path today. I removed it because I wasn't sure on the intention of a "processing server" as a concept vs the dispatcher. I'll consolidate, it was cleaner that way, thanks.

**Comment by matt-psaltis** [comment]
> Hello @matt-psaltis ,
> 
> Since the object that implements `IProcessingServer` will be started in `Bootstrapper` class so what do you think about making the `IDispatcher` inherit the `IProcessingServer` ?
> 
> https://github.com/dotnetcore/CAP/blob/master/src/DotNetCore.CAP/CAP.ServiceCollectionExtensions.cs#L47
> https://github.com/dotnetcore/CAP/blob/master/src/DotNetCore.CAP/Internal/IBootstrapper.Default.cs#L33

I remember why I unwound this now, it was because of the number of places that directly reference the IDispatcher.

I could register it as an IProcessingServer using something like this:
```
            services.TryAddEnumerable(ServiceDescriptor.Singleton<IProcessingServer, IDispatcher>(sp =>
                sp.GetRequiredService<IDispatcher>()));
```

Need to make sure we only get a single instance created so by registering it as an IProcessing Server as well as IDispatcher, we need to use a factory pattern like above to reference the same IDispatcher instance. Does that make sense?
Happy to do it this way - I'm just being overly cautious of making drastic changes like I would with my own repositories :D

**Comment by yang-xiaodong** [comment]
It’s a good idea to use a factory to resolve, just do it

**Comment by matt-psaltis** [comment]
> It’s a good idea to use a factory to resolve, just do it

Done.

**Review by yang-xiaodong — APPROVED** [review]

**Comment by yang-xiaodong** [comment]
Looks good, thanks very much

**Pull request merged**